src/main/java/uk/gov/di/ipv/cri/common/library/service/SessionService.java

```diff
@@ -84,16 +89,69 @@ public SessionItem getSession(String sessionId) {
         return dataStore.getItem(sessionId);
     }
 
-    public SessionItem getSessionByAccessToken(AccessToken accessToken) {
-        return listUtil.getOneItemOrThrowError(
-                dataStore.getItemByIndex(
-                        SessionItem.ACCESS_TOKEN_INDEX, accessToken.toAuthorizationHeader()));
+    public SessionItem getSessionByAccessToken(AccessToken accessToken)
+            throws SessionExpiredException, AccessTokenExpiredException, SessionNotFoundException {
+        SessionItem sessionItem;
+
+        try {
+            sessionItem =
+                    listUtil.getOneItemOrThrowError(
+                            dataStore.getItemByIndex(
+                                    SessionItem.ACCESS_TOKEN_INDEX,
+                                    accessToken.toAuthorizationHeader()));
+        } catch (IllegalArgumentException e) {
+            if (e.getMessage().contains("No items found")) {
+                throw new SessionNotFoundException("no session found with that access token");
+            } else {
+                throw new SessionNotFoundException(
+                        "more than one session found with that access token");
+            }
+        }
+
+        // Re-fetch our session directly to avoid problems with projections
+        sessionItem = getSession(String.valueOf(sessionItem.getSessionId()));
```

Comment:
why is the session item retrieved a second time here?

Reply:
Added due to the fields not (yet) being in the projection - will raise future PRs to correct the projections